linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/FunctionEmitter.scala

```diff
@@ -2217,17 +2217,26 @@ private[emitter] class FunctionEmitter(sjsGen: SJSGen) {
 
         case Apply(_, receiver, method, args) =>
           val methodName = method.name
-          val newReceiver = transformExprNoChar(receiver)
+
+          def newReceiver(asChar: Boolean): js.Tree = {
+            if (!asChar)
+              transformExprNoChar(receiver)
+            else if (realTreeType(receiver) == CharType)
+              transformExpr(receiver, preserveChar = true)
+            else
+              transformExpr(AsInstanceOf(receiver, CharType), preserveChar = true)
```

Comment:
Just to check my understanding: This is necessary, because CharType is the only type where non-null values of the boxed type are not genuine values of the primitive type. So an actual conversion needs to happen.

Again, just like in the optimizer, we need not care about nulls, because that is UB (in the context of a function call). Otherwise we'd need to unbox for all primitive types, since they'd need to convert null to the zero.

Reply:
Exactly. I have reordered and added comments here to make that clear. Also for your other comment above.